### PR TITLE
cli: write infrastructure to new state file

### PR DESCRIPTION
**Pull request by elchead**
### Context

This is the first implementation part of the new constellation state (#2281).

### Proposed change(s)
create
- aggregate infrastructure state from TF output
- write the infrastructure output to  `constellation-state.yaml`  with version v1 after terraform apply

upgrade
- extend TF client with ShowInfrastrucutre() to output the infrastructure part of the state
- use ShowInfrastructure() and overwrite the infrastructure output to  `constellation-state.yaml`  with version v1 after terraform apply

### Related issue
- [AB#3422](https://dev.azure.com/Edgeless/ae37573d-ccde-4af2-ab1e-001e587197d1/_workitems/edit/3422)

### Checklist

- [ ] Update [docs](https://github.com/edgelesssys/constellation/tree/main/docs)
- [x] Add labels (e.g., for changelog category)
- [ ] Is PR title adequate for changelog?
- [x] Link to Milestone

### CODE REVIEW

File: cli/internal/state/state.go
Comment (derpsteb) [inline cli/internal/state/state.go:23]: It seems more idiomatic to me to make this an interface and add marshalling methods. Would prevent setting both keys, makes nil checks obsolete, makes the intent of the property clearer. Was there a specific reason you chose this approach, apart from the fact that this is how we do it in the config?

**Comment by derpsteb** [comment]
LGTM. Would just like to discuss that one point.

File: cli/internal/state/state.go
Comment (elchead) [reply cli/internal/state/state.go:23]: you mean to make this field contain a CSP interface similar to the attestation config?
In the end the properties of the CSPs are very different so I don't see how to generalize the fields in an interface.
What is unclear about the intent of the property?

Do you have a suggestion?

Comment (derpsteb) [reply cli/internal/state/state.go:23]: True. It isn't clear how to write that interface.

**Review by derpsteb — APPROVED** [review]

**Comment by derpsteb** [comment]
Before merging this: what is the plan to ensure that the migration from the idfile to the statefile happens within 1 release cycle? How do we migrate?

**Comment by elchead** [comment]
> Before merging this: what is the plan to ensure that the migration from the idfile to the statefile happens within 1 release cycle? How do we migrate?

The idFile is only removed in the last step (see user stories in Azure). So the code is not broken at any point.
When we prioritize this story, it should be very feasible to merge this in the next release.
Otherwise, we would have both the idFile and state file during the intermediate release.

**Comment by derpsteb** [comment]
Good to merge from my side.

File: cli/internal/cmd/create.go
Comment (daniel-weisse) [inline cli/internal/cmd/create.go:180]: This should be a constant.
See the version number of the config file for reference

File: cli/internal/cmd/upgradeapply.go
Comment (daniel-weisse) [inline cli/internal/cmd/upgradeapply.go:206]: Same as before

File: cli/internal/cmd/upgradeapply.go
Comment (daniel-weisse) [inline cli/internal/cmd/upgradeapply.go:211]: Consider moving the updating of an existing state file to `updateClusterIDFile` (rename the function eventually),
and have an explicit migration function if the state file does not exist when running `upgrade apply`.
IMO does not need to happen in this PR, but should be done before this is in any release.
If you decide to implement in a different PR, please add a comment to this section.

File: cli/internal/state/state.go
Comment (daniel-weisse) [inline cli/internal/state/state.go:20]: I'm unsure if `InitSecret` contains potentially unprintable characters.
It was of type `[]byte` in the id file, maybe we should keep it as such here?
@3u13r can you give an opinion on this?

File: cli/internal/state/state.go
Comment (daniel-weisse) [inline cli/internal/state/state.go:1]: Can we add a Codeowner for this package?

Comment (daniel-weisse) [inline cli/internal/state/state.go:1]: Doc comments in this file need some work, but should probably be handled in [AB#3426](https://dev.azure.com/Edgeless/Edgeless/_workitems/edit/3426)

File: cli/internal/terraform/terraform.go
Comment (daniel-weisse) [inline cli/internal/terraform/terraform.go:175]: Is there still a reason to keep the old `ShowCluster` (and also the `CreateCluster`) function which returns `ApplyOutput`, instead of letting the terraform package just return `state.Infrastructure`?

**Review by daniel-weisse — CHANGES_REQUESTED** [review]

File: cli/internal/terraform/terraform.go
Comment (daniel-weisse) [reply cli/internal/terraform/terraform.go:175]: Nvm, there is already a PR for that

File: cli/internal/state/state.go
Comment (elchead) [reply cli/internal/state/state.go:1]: Sure, who should be it? I would opt in since i don't have so many yet. Or do you want it?

File: cli/internal/state/state.go
Comment (3u13r) [reply cli/internal/state/state.go:20]: Going for `string` is fine here. The `initSecret` is generated in terraform (e.g. https://github.com/edgelesssys/constellation/blob/main/cli/internal/terraform/terraform/gcp/main.tf#L69) and in terraform `random_password` is equivalent to `random_string` which only consists of printable characters.

File: cli/internal/state/state.go
Comment (daniel-weisse) [reply cli/internal/state/state.go:1]: Sure, go ahead.

**Review by daniel-weisse — APPROVED** [review]

**Review by 3u13r — DISMISSED** [review]

**Review by 3u13r — APPROVED** [review]
LGTM

**Pull request merged**
